metricbeat/module/kafka/partition/partition.go

```diff
@@ -19,75 +21,196 @@ func init() {
 // MetricSet type defines all fields of the partition MetricSet
 type MetricSet struct {
 	mb.BaseMetricSet
-	client sarama.Client
+
+	broker *sarama.Broker
+	cfg    *sarama.Config
+	id     int32
 }
 
-// New creates a new instance of the partition MetricSet
+var noID int32 = -1
+
+var errFailQueryOffset = errors.New("Failed to query offset for")
+
+// New create a new instance of the partition MetricSet
 func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
-	logp.Warn("EXPERIMENTAL: The %v %v metricset is experimental", base.Module().Name(), base.Name())
+	config := struct{}{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	cfg := sarama.NewConfig()
+	cfg.Net.DialTimeout = base.Module().Config().Timeout
+	cfg.Net.ReadTimeout = base.Module().Config().Timeout
+	cfg.ClientID = "metricbeat"
+
+	broker := sarama.NewBroker(base.Host())
+	return &MetricSet{
+		BaseMetricSet: base,
+		broker:        broker,
+		cfg:           cfg,
+		id:            noID,
+	}, nil
+}
+
+func (m *MetricSet) connect() (*sarama.Broker, error) {
+	b := m.broker
+	if err := b.Open(m.cfg); err != nil {
+		return nil, err
+	}
+
+	if m.id != noID {
+		return b, nil
+	}
 
-	return &MetricSet{BaseMetricSet: base}, nil
+	// current broker is bootstrap only. Get metadata to find id:
+	meta, err := b.GetMetadata(&sarama.MetadataRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	addr := b.Addr()
+	for _, other := range meta.Brokers {
+		if other.Addr() == addr {
+			m.id = other.ID()
+			break
+		}
+	}
+
+	if m.id == noID {
+		b.Close()
+		err = fmt.Errorf("No advertised broker with address %v found", addr)
+		return nil, err
+	}
+
+	return b, nil
 }
 
 // Fetch partition stats list from kafka
 func (m *MetricSet) Fetch() ([]common.MapStr, error) {
-	if m.client == nil {
-		config := sarama.NewConfig()
-		config.Net.DialTimeout = m.Module().Config().Timeout
-		config.Net.ReadTimeout = m.Module().Config().Timeout
-		config.ClientID = "metricbeat"
-
-		client, err := sarama.NewClient([]string{m.Host()}, config)
-		if err != nil {
-			return nil, err
-		}
-		m.client = client
+	b, err := m.connect()
+	if err != nil {
+		return nil, err
 	}
 
-	topics, err := m.client.Topics()
+	defer b.Close()
+	response, err := b.GetMetadata(&sarama.MetadataRequest{})
```

Comment:
might fail if leader-election is active? consider retry (max 3 times) with waits of a few milliseconds (see sarama settings for defaults)